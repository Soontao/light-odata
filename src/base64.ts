/* eslint-disable max-len */
/* eslint-disable prefer-const */
// Array of bytes to Base64 string decoding
function b64ToUint6(nChr) {
  return nChr > 64 && nChr < 91 ?
    nChr - 65
    : nChr > 96 && nChr < 123 ?
      nChr - 71
      : nChr > 47 && nChr < 58 ?
        nChr + 4
        : nChr === 43 ?
          62
          : nChr === 47 ?
            63
            :
            0;
}

function base64DecToArr(sBase64: string, nBlocksSize?: number) {
  const sB64Enc = sBase64.replace(/[^A-Za-z0-9+/]/g, "");
  const nInLen = sB64Enc.length;
  const nOutLen = nBlocksSize ? Math.ceil((nInLen * 3 + 1 >> 2) / nBlocksSize) * nBlocksSize : nInLen * 3 + 1 >> 2;
  const taBytes = new Uint8Array(nOutLen);

  let nMod3;
  let nMod4;
  let nUint24 = 0;
  let nOutIdx = 0;
  for (let nInIdx = 0; nInIdx < nInLen; nInIdx++) {
    nMod4 = nInIdx & 3;
    nUint24 |= b64ToUint6(sB64Enc.charCodeAt(nInIdx)) << 6 * (3 - nMod4);
    if (nMod4 === 3 || nInLen - nInIdx === 1) {
      nMod3 = 0;
      while (nMod3 < 3 && nOutIdx < nOutLen) {
        taBytes[nOutIdx] = nUint24 >>> (16 >>> nMod3 & 24) & 255;
        nMod3++;
        nOutIdx++;
      }
      nUint24 = 0;

    }
  }

  return taBytes;
}

/* Base64 string to array encoding */
function uint6ToB64(nUint6: number) {
  return nUint6 < 26 ?
    nUint6 + 65
    : nUint6 < 52 ?
      nUint6 + 71
      : nUint6 < 62 ?
        nUint6 - 4
        : nUint6 === 62 ?
          43
          : nUint6 === 63 ?
            47
            :
            65;
}

function base64EncArr(aBytes) {
  let nMod3 = 2;
  let sB64Enc = "";

  const nLen = aBytes.length;
  let nUint24 = 0;
  for (let nIdx = 0; nIdx < nLen; nIdx++) {
    nMod3 = nIdx % 3;
    if (nIdx > 0 && (nIdx * 4 / 3) % 76 === 0) {
      sB64Enc += "\r\n";
    }

    nUint24 |= aBytes[nIdx] << (16 >>> nMod3 & 24);
    if (nMod3 === 2 || aBytes.length - nIdx === 1) {
      sB64Enc += String.fromCodePoint(uint6ToB64(nUint24 >>> 18 & 63), uint6ToB64(nUint24 >>> 12 & 63), uint6ToB64(nUint24 >>> 6 & 63), uint6ToB64(nUint24 & 63));
      nUint24 = 0;
    }
  }
  return sB64Enc.substring(0, sB64Enc.length - 2 + nMod3) + (nMod3 === 2 ? "" : nMod3 === 1 ? "=" : "==");
}

/* UTF-8 array to JS string and vice versa */

function utf8ArrToStr(aBytes) {
  let sView = "";
  let nPart;
  const nLen = aBytes.length;
  for (let nIdx = 0; nIdx < nLen; nIdx++) {
    nPart = aBytes[nIdx];
    sView += String.fromCodePoint(
      nPart > 251 && nPart < 254 && nIdx + 5 < nLen ? /* six bytes */
        /* (nPart - 252 << 30) may be not so safe in ECMAScript! So…: */
        (nPart - 252) * 1073741824 + (aBytes[++nIdx] - 128 << 24) + (aBytes[++nIdx] - 128 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
        : nPart > 247 && nPart < 252 && nIdx + 4 < nLen ? /* five bytes */
          (nPart - 248 << 24) + (aBytes[++nIdx] - 128 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
          : nPart > 239 && nPart < 248 && nIdx + 3 < nLen ? /* four bytes */
            (nPart - 240 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
            : nPart > 223 && nPart < 240 && nIdx + 2 < nLen ? /* three bytes */
              (nPart - 224 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
              : nPart > 191 && nPart < 224 && nIdx + 1 < nLen ? /* two bytes */
                (nPart - 192 << 6) + aBytes[++nIdx] - 128
                : /* nPart < 127 ? */ /* one byte */
                nPart
    );
  }
  return sView;
}

function strToUTF8Arr(sDOMStr: string): Uint8Array {
  let aBytes;
  let nChr;
  const nStrLen = sDOMStr.length;
  let nArrLen = 0;

  /* mapping… */
  for (let nMapIdx = 0; nMapIdx < nStrLen; nMapIdx++) {
    nChr = sDOMStr.codePointAt(nMapIdx);

    if (nChr > 65536) {
      nMapIdx++;
    }

    nArrLen += nChr < 0x80 ? 1 : nChr < 0x800 ? 2 : nChr < 0x10000 ? 3 : nChr < 0x200000 ? 4 : nChr < 0x4000000 ? 5 : 6;
  }

  aBytes = new Uint8Array(nArrLen);

  /* transcription… */
  let nIdx = 0;
  let nChrIdx = 0;
  while (nIdx < nArrLen) {
    nChr = sDOMStr.codePointAt(nChrIdx);
    if (nChr < 128) {
      /* one byte */
      aBytes[nIdx++] = nChr;
    } else if (nChr < 0x800) {
      /* two bytes */
      aBytes[nIdx++] = 192 + (nChr >>> 6);
      aBytes[nIdx++] = 128 + (nChr & 63);
    } else if (nChr < 0x10000) {
      /* three bytes */
      aBytes[nIdx++] = 224 + (nChr >>> 12);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
    } else if (nChr < 0x200000) {
      /* four bytes */
      aBytes[nIdx++] = 240 + (nChr >>> 18);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    } else if (nChr < 0x4000000) {
      /* five bytes */
      aBytes[nIdx++] = 248 + (nChr >>> 24);
      aBytes[nIdx++] = 128 + (nChr >>> 18 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    } else /* if (nChr <= 0x7fffffff) */ {
      /* six bytes */
      aBytes[nIdx++] = 252 + (nChr >>> 30);
      aBytes[nIdx++] = 128 + (nChr >>> 24 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 18 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    }
    nChrIdx++;
  }

  return aBytes;
}

const encodeJs = (input: string) => {
  return base64EncArr(strToUTF8Arr(input));
};

const decodeJs = (input: string) => {
  return utf8ArrToStr(base64DecToArr(input));
};

/**
 * encode string to base64
 *
 * @param input
 */
const encode = function (input = ""): string {
  if (globalThis.Buffer !== undefined) {
    return globalThis.Buffer.from(input, "utf-8").toString("base64");
  }
  return base64EncArr(strToUTF8Arr(input));
};

/**
 * decode base64 to string
 *
 * @param input base64 string
 */
const decode = function (input: string): string {
  if (globalThis.Buffer !== undefined) {
    return globalThis.Buffer.from(input, "base64").toString("utf-8");
  }
  return utf8ArrToStr(base64DecToArr(input));
};

export { encode, decode, encodeJs, decodeJs };
