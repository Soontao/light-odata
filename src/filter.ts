/* eslint-disable max-len */
import { join } from "@newdash/newdash/join";
import { FrameworkError, ValidationError } from "./errors";
import { SystemQueryOptions } from "./params";
import { ODataDateTime, ODataDateTimeOffset, ODataValueObject } from "./types";


export type FilterValue = number | string | ODataValueObject | null | boolean | Symbol | object

export enum ExprOperator {
  eq = "eq",
  ne = "ne",
  gt = "gt",
  lt = "lt",
  ge = "ge",
  le = "le",
}

type FieldExpr = {
  op: ExprOperator;
  value: FilterValue;
}

type FieldExprMappings = {
  [key: string]: FieldExpr[]
}

/**
 * odata property filter expression
 * @private
 * @internal
 */
class ODataPropertyExpr<T extends ODataFilter> {

  constructor(filter: T, fieldName: string, mapping: FieldExprMappings) {
    this._exprMappings = mapping;
    this._fieldName = fieldName;
    this._filter = filter;
    // initilize
    if (this._getFieldExprs() == undefined) {
      this._exprMappings[this._fieldName] = [];
    }
  }

  private _filter: T;

  private _fieldName: string;

  private _exprMappings: FieldExprMappings;

  private _getFieldExprs() {
    return this._exprMappings[this._fieldName];
  }

  private _addExpr(op: ExprOperator, value: FilterValue) {

    switch (typeof value) {
      case "number": case "boolean":
        this._getFieldExprs().push({ op, value: `${value}` });
        break;
      case "string":
        if (value.startsWith("'") || value.startsWith("datetime")) {
          this._getFieldExprs().push({ op, value });
        } else {
          this._getFieldExprs().push({ op, value: `'${value}'` });
        }
        break;
      case "object":
        if (value instanceof ODataValueObject) {
          this._getFieldExprs().push({ op, value: value.toString() });
        } else if (value === null) {
          this._getFieldExprs().push({ op, value: null });
        } else {
          throw new FrameworkError(`Not support object ${value?.["constructor"]?.["name"] || typeof value} in odata filter eq/ne/gt/ge/ne/nt ...`);
        }
        break;
      case "undefined":
        throw new ValidationError(`You must set value in odata filter eq/ne/gt/ge/ne/nt ...`);
      default:
        throw new FrameworkError(`Not support typeof ${typeof value}: ${value} in odata filter eq/ne/gt/ge/ne/nt ...`);
    }

  }

  /**
   * equal
   * @param value
   */
  eq(value: FilterValue): T {
    this._addExpr(ExprOperator.eq, value);
    return this._filter;
  }

  /**
   * not equal
   * @param value
   */
  ne(value: FilterValue): T {
    this._addExpr(ExprOperator.ne, value);
    return this._filter;
  }

  /**
   * @deprecated use `filter.eq("value")` instead
   * @param value
   * @returns
   */
  eqString(value: string): T {
    this._addExpr(ExprOperator.eq, `'${value}'`);
    return this._filter;
  }

  /**
   * @deprecated use `filter.ne("value")` instead
   * @param value
   * @returns
   */
  neString(value: string): T {
    this._addExpr(ExprOperator.ne, `'${value}'`);
    return this._filter;
  }

  /**
   * greater or equal
   * @param value
   */
  ge(value: FilterValue): T {
    this._addExpr(ExprOperator.ge, value);
    return this._filter;
  }

  /**
   * greater than
   * @param value
   */
  gt(value: FilterValue): T {
    this._addExpr(ExprOperator.gt, value);
    return this._filter;
  }

  /**
   * less or equal
   * @param value
   */
  le(value: FilterValue): T {
    this._addExpr(ExprOperator.le, value);
    return this._filter;
  }

  /**
   * less than
   * @param value
   */
  lt(value: FilterValue): T {
    this._addExpr(ExprOperator.lt, value);
    return this._filter;
  }

  /**
   * match any value in an array
   *
   * @param values
   */
  in(values: FilterValue[] = []): T {
    if (values.length > 0) {
      values.forEach((value) => {
        this.eq(value);
      });
    }
    return this._filter;
  }

  /**
   * filter by value range
   *
   * @param low
   * @param max
   * @param includeBoundary
   */
  between(low: FilterValue, max: FilterValue, includeBoundary = true): T {
    if (low == undefined || max == undefined) {
      throw new ValidationError("You must give out the start and end value");
    }
    if (includeBoundary) {
      this.ge(low);
      this.le(max);
    } else {
      this.gt(low);
      this.lt(max);
    }
    return this._filter;
  }

  /**
   *
   * @deprecated please use filter.between(ODataDateTime.from(date), ODataDateTime.from(date))
   * @param start
   * @param end
   * @param includeBoundary
   * @returns
   */
  betweenDateTime(start?: Date, end?: Date, includeBoundary = true): T {
    if (start == undefined && end == undefined) {
      throw new ValidationError("You must give out the start or end date");
    }
    if (start instanceof Date) {
      if (includeBoundary) {
        this.ge(ODataDateTime.from(start));
      } else {
        this.gt(ODataDateTime.from(start));
      }
    }
    if (end instanceof Date) {
      if (includeBoundary) {
        this.le(ODataDateTime.from(end));
      } else {
        this.lt(ODataDateTime.from(end));
      }
    }
    return this._filter;
  }

  /**
   *  @deprecated please use filter.between(ODataDateTimeOffset.from(date), ODataDateTimeOffset.from(date))
   * @param start
   * @param end
   * @param includeBoundary
   * @returns
   */
  betweenDateTimeOffset(start?: Date, end?: Date, includeBoundary = true): T {
    if (start == undefined && end == undefined) {
      throw new ValidationError("You must give out the start or end date");
    }
    if (start instanceof Date) {
      if (includeBoundary) {
        this.ge(ODataDateTimeOffset.from(start));
      } else {
        this.gt(ODataDateTimeOffset.from(start));
      }
    }
    if (end instanceof Date) {
      if (includeBoundary) {
        this.le(ODataDateTimeOffset.from(end));
      } else {
        this.lt(ODataDateTimeOffset.from(end));
      }
    }
    return this._filter;
  }

}


/**
 * OData filter builder
 *
 */
export class ODataFilter<T = any> {

  static newBuilder<T = any>() {
    return new ODataFilter<T>();
  }

  /**
   * construct a new filter
   */
  static newFilter<T = any>() {
    return new ODataFilter<T>();
  }

  private _fieldExprMappings: FieldExprMappings = {};

  /**
   * getExprMapping
   *
   * @internal
   * @private
   */
  private getExprMapping(): FieldExprMappings {
    return this._fieldExprMappings;
  }

  /**
   * filter property
   * @see property
   * @param name field name or function expression
   */
  field(name: keyof T): ODataPropertyExpr<this> {
    return this.property(name);
  }

  /**
   * filter property
   *
   * @param name property name
   */
  property(name: keyof T): ODataPropertyExpr<this> {
    // @ts-ignore
    return new ODataPropertyExpr(this, name, this.getExprMapping());
  }

  /**
   * The value of a field matches any value in the list.
   *
   * @deprecated please use filter.field().in()
   * @param name
   * @param values
   */
  fieldIn(name: keyof T, values: string[]): this {
    // @ts-ignore
    return this.fieldValueMatchArray(name, values);
  }

  /**
   * The value of a field matches any value in the list.
   *
   * @deprecated please use filter.field().in()
   * @param name
   * @param values
   */
  fieldValueMatchArray(name: keyof T, values: string[] = []): this {
    if (values) {
      values.forEach((value) => {
        this.field(name).eqString(value);
      });
    }
    return this;
  }

  /**
   * DEPRECATED
   *
   * please use betweenDateTime/betweenDateTimeOffset
   *
   * @deprecated
   * @param name
   * @param start
   * @param end
   */
  inPeriod(name: keyof T, start: Date, end: Date): this {
    return this.betweenDateTime(name, start, end);
  }

  /**
   * @deprecated
   * @param name
   * @param start
   * @param end
   */
  betweenDateTime(name: keyof T, start: Date, end: Date): this {
    if (start && end) {
      return this.gtDateTime(name, start).ltDateTime(name, end);
    }
    throw new ValidationError("You must give out the start and end date");

  }

  /**
   * @deprecated
   * @param name
   * @param start
   * @param end
   */
  betweenDateTimeOffset(name: keyof T, start: Date, end: Date): ODataFilter {
    if (start && end) {
      return this.gtDateTimeOffset(name, start).ltDateTimeOffset(name, end);
    }
    throw new ValidationError("You must give out the start and end date");
  }

  /**
   * @deprecated
   * @param name
   * @param date
   */
  gtDateTime(name: keyof T, date: Date): this {
    return this.field(name).gt(ODataDateTime.from(date));
  }

  /**
   * @deprecated
   * @param name
   * @param date
   */
  gtDateTimeOffset(name: keyof T, date: Date): this {
    return this.field(name).gt(ODataDateTimeOffset.from(date));
  }

  /**
   * @deprecated
   * @param name
   * @param date
   */
  ltDateTime(name: keyof T, date: Date): this {
    return this.field(name).lt(ODataDateTime.from(date));
  }

  /**
   * @deprecated
   * @param name
   * @param date
   */
  ltDateTimeOffset(name: keyof T, date: Date): this {
    return this.field(name).lt(ODataDateTimeOffset.from(date));
  }

  /**
   * @deprecated c4codata will auto group exprs
   * @param filter
   */
  group(filter: ODataFilter): this {
    this._fieldExprMappings = Object.assign(this._fieldExprMappings, filter.getExprMapping());
    return this;
  }

  toString(): string {
    return this.build();
  }

  protected _buildFieldExprString(field: string): string {
    const exprs = this.getExprMapping()[field];
    if (exprs.length > 0) {
      if (exprs.filter((expr) => expr.op == ExprOperator.eq).length == 0) {
        return `(${join(
          exprs.map(({ op, value }) => `${field} ${op} ${value}`),
          " and "
        )})`;
      }
      return `(${join(
        exprs.map(({ op, value }) => `${field} ${op} ${value}`),
        " or "
      )})`;

    }
    return "";


  }

  /**
   * build filter to string
   */
  build(): string {
    let _rt = "";
    _rt = join(
      // join all fields exprs string
      Object.entries(this.getExprMapping()).map(
        ([fieldName, exprs]) => {
          switch (exprs.length) {
            // if one field expr mapping array is empty
            case 0:
              return "";
            // only have one expr
            case 1:
              const { op, value } = exprs[0];
              return `${fieldName} ${op} ${value}`;
            default:
              // multi exprs
              return this._buildFieldExprString(fieldName);
          }

        }
      ),
      " and "
    );
    return _rt;
  }

}

export class ParamBoundedFilter<T = any> extends ODataFilter<T> {

  #params: SystemQueryOptions;

  constructor(params: SystemQueryOptions) {
    super();
    this.#params = params;
  }

  /**
   * the filter is ended, return the param object
   *
   * @returns the param
   */
  public filterEnd(): SystemQueryOptions {
    return this.#params.filter(this);
  }

}

export function filter() {
  return ODataFilter.newFilter();
}
