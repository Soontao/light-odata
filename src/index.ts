import { OData } from "./odata";

export * from "./batch";
export * from "./filter";
export * from "./params";
export * from "./types";
export { EdmV4 } from "./types_v4";
export * from "./util";
export * from "./request";
export { transformation } from "./tranformation";
export { OData };

export default OData;
