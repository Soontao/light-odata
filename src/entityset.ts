// @ts-nocheck
import { ODataServerError } from "./errors";
import { ODataFilter } from "./filter";
import { systemOptions, SystemQueryOptions } from "./params";
import { OData } from "./odata";
import type { DeepPartial, ODataKeyPredicate, PlainODataResponse } from "./types";

/**
 * EntitySet
 */
export class EntitySet<T = any> {

  private _entitySet: string;

  private _client: OData;

  constructor(collection: string, client: OData) {
    this._entitySet = collection;
    this._client = client;
  }

  private _checkError(res: any): void {
    if (res.error) {
      switch (this._client.getVersion()) {
        case "v2":
          throw new ODataServerError(res.error?.message?.value);
        case "v4":
          throw new ODataServerError(res.error?.message);
        default:
          break;
      }
    }
  }

  private _getResultSingle(res: PlainODataResponse) {
    switch (this._client.getVersion()) {
      case "v2":
        // @ts-ignore
        return res.d;
      case "v4":
        // @ts-ignore
        return res;
      default:
        break;
    }
  }

  private _getResults(res: PlainODataResponse) {
    switch (this._client.getVersion()) {
      case "v2":
        // @ts-ignore
        return res.d?.results;
      case "v4":
        // @ts-ignore
        return res?.value;
      default:
        break;
    }
  }

  public newFilter(): ODataFilter<T> {
    return this._client.newFilter();
  }

  /**
   * @deprecated
   * @returns
   */
  public newParam(): SystemQueryOptions<T> {
    return this._client.newParam();
  }

  public newOptions() {
    return this._client.newOptions();
  }

  async retrieve(id: ODataKeyPredicate, systemOptions?: SystemQueryOptions): Promise<T> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      id,
      params: systemOptions
    });
    this._checkError(res);
    return this._getResultSingle(res);
  }


  async find(base: DeepPartial<T>): Promise<T[]> {
    const filter = OData.newFilter();

    Object.entries(base).forEach(([key, value]) => {
      if (typeof value == "string") {
        filter.field(key).eqString(value);
      } else {
        filter.field(key).eq(value);
      }
    });

    return this.query(OData.newParam().filter(filter));
  }

  async query(filter: ODataFilter): Promise<T[]>;

  async query(systemOptions: SystemQueryOptions): Promise<T[]>;

  async query(): Promise<T[]>;

  async query(param: any): Promise<any> {
    if (param instanceof ODataFilter) {
      param = systemOptions().filter(param);
    }
    if (param == undefined) {
      param = OData.newParam();
    }
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      params: param
    });
    this._checkError(res);
    return this._getResults(res);
  }

  async count(filter?: ODataFilter): Promise<number>;

  async count(filter?: Partial<T>): Promise<number>;

  async count(filter?: any) {

    const params = OData.newParam().top(1);

    // set count flag
    switch (this._client.getVersion()) {
      case "v4":
        params.count(true);
        break;
      case "v2":
        params.inlinecount(true);
        break;
    }

    if (typeof filter === "object") {

      if (filter instanceof ODataFilter) {
        params.filter(filter);
      } else {
        const newFilter = this.newFilter();
        for (const [key, value] of Object.entries(filter)) {
          newFilter.field(key).eq(value);
        }
        params.filter(newFilter);
      }

    }

    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      params
    });

    this._checkError(res);
    switch (this._client.getVersion()) {
      case "v2":
        return parseInt(res?.d?.__count);
      case "v4":
        return res["@odata.count"];
    }
  }

  async create(body: DeepPartial<T>): Promise<T>;

  async create(body: any): Promise<T>;

  async create(body: DeepPartial<T>): Promise<T> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "POST",
      entity: body
    });
    this._checkError(res);
    return this._getResultSingle(res);

  }

  /**
   * update instance by `UPDATE` method
   *
   * @param id
   * @param body
   */
  async update(id: ODataKeyPredicate, body: DeepPartial<T>): Promise<void> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "PATCH",
      id,
      entity: body
    });
    this._checkError(res);
  }

  /**
   * delete instance by object keys
   *
   * @param id
   */
  async delete(id: ODataKeyPredicate): Promise<void> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "DELETE",
      id
    });
    this._checkError(res);
  }

  /**
   * perform bounded action
   *
   * @param actionName the action name, remember add namespace for it
   * @param parameters
   */
  async action(actionName: string, id: ODataKeyPredicate, parameters?: any): any {
    const responseBody = await this._client.newRequest({
      collection: this._entitySet,
      method: "POST",
      id,
      parameters,
      actionName
    });
    this._checkError(responseBody);
    return responseBody;
  }

  /**
   * perform bounded function
   *
   * @param functionName
   * @param id
   */
  async function(
    functionName: string,
    id: ODataKeyPredicate,
    parameters?: any,
    systemOptions?: SystemQueryOptions
  ): any {
    const responseBody = await this._client.newRequest({
      collection: this._entitySet,
      method: "GET",
      id,
      params: systemOptions,
      parameters,
      functionName
    });
    this._checkError(responseBody);
    return responseBody;
  }



}
