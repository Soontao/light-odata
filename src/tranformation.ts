import { ODataFilter } from "./filter";


/**
 * Transformation for `$apply`
 *
 * @experimental
 */
export class Transformation {

  private _aggregate: string;

  private _countSumPercent: {
    count: number;
    valueExpression: string;
    type: "sum" | "count" | "percent";
    direction: "top" | "bottom";
  };

  private _identity: boolean;

  private _concat: Transformation[];

  private _groupBy: {
    properties: Array<string>;
    transformations: Transformation[];
  };

  private _filter: ODataFilter<any>;

  public static newTransformation() {
    return new Transformation();
  }

  private constructor() { }

  /**
   * Aggregate expressions can define an alias using the as keyword,
   * followed by a SimpleIdentifier
   *
   * @see [Transformation aggregate](http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs01/odata-data-aggregation-ext-v4.0-cs01.html#_The_aggregate_Transformation_1)
   * @param expression
   * @experimental
   * @example
   *
   * ```js
   * aggregate('Amount with sum as Total')
   * ```
   * @returns
   */
  public aggregate(expression: string) {
    this._aggregate = expression;
    return this;
  }

  /**
   *
   * @param count The first parameter specifies the number of instances to return in the transformed set. currently only support number
   * @param valueExpression The second parameter specifies the value by which the instances are compared for determining the result set.
   */
  public topcount(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "top",
      type: "count"
    };
    return this;
  }

  public topsum(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "top",
      type: "sum"
    };
    return this;
  }

  public toppercent(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "top",
      type: "percent",
    };
    return this;
  }


  public bottomcount(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "bottom",
      type: "count"
    };
    return this;
  }

  public bottomsum(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "bottom",
      type: "sum"
    };
    return this;
  }

  public bottompercent(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "bottom",
      type: "percent",
    };
    return this;
  }

  /**
   * The identity transformation returns its input set.
   *
   * @returns
   */
  public identity() {
    this._identity = true;
    return this;
  }

  /**
   * The concat transformation takes two or more parameters,
   * each of which is a sequence of set transformations.
   *
   * @param transformations
   * @returns
   */
  public concat(...transformations: Array<Transformation>) {
    this._concat = transformations;
    return this;
  }

  /**
   * The filter transformation takes a Boolean expression that could also be passed as a `$filter` system query option to its input set and returns all instances for which this expression evaluates to true.
   * @param filter
   * @returns
   */
  public filter(filter: ODataFilter) {
    this._filter = filter;
    return this;
  }

  /**
   *
    The `groupby` transformation takes one or two parameters and
    1. Splits the initial set into subsets where all instances in a subset have the same values for the grouping properties specified in the first parameter,
    2. Applies set transformations to each subset according to the second parameter, resulting in a new set of potentially different structure and cardinality,
    3. Ensures that the instances in the result set contain all grouping properties with the correct values for the group,
    4. Concatenates the intermediate result sets into one result set.

   * @see [Transformation groupby](http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs01/odata-data-aggregation-ext-v4.0-cs01.html#_Toc378326304)
   * @param properties
   * @param transformations
   * @experimental
   * @returns
   */
  public groupBy(
    properties: Array<string>,
    ...transformations: Array<Transformation>
  ) {

    // TODO: properties can not be empty
    this._groupBy = {
      properties,
      transformations
    };
    return this;
  }

  // TODO: filter, expand, search, isdefined, $crossjoin


  public toString() {
    const expressionText = "";
    if (this._aggregate !== undefined) {
      return `aggregate(${this._aggregate})`;
    }

    if (this._countSumPercent !== undefined) {
      const { direction, type, count, valueExpression } = this._countSumPercent;
      return `${direction}${type}(${count},${valueExpression})`;
    }

    if (this._groupBy !== undefined) {
      const properties = `(${this._groupBy.properties.join(",")})`;
      if (this._groupBy.transformations.length > 0) {
        const transformations = this
          ._groupBy
          .transformations
          .map(t => t.toString())
          .join("/");
        // see Transformation Sequences
        return `groupBy(${properties},${transformations})`;
      }
      return `groupBy(${properties})`;
    }

    if (this._identity === true) {
      return "identity";
    }

    if (this._filter !== undefined) {
      //  $apply is only avaible for odata v4,
      return `filter(${this._filter.toString()})`;
    }

    // TODO: other transformation

    return expressionText;
  }

}


export function transformation() {
  return Transformation.newTransformation();
}
