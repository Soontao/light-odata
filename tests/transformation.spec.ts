import { ODataFilter } from "../src";
import { transformation, Transformation } from "../src/tranformation";


describe('Transofmration Test Suite', () => {

  it('should support to create instnace', () => {
    expect(Transformation.newTransformation()).toBeInstanceOf(Transformation)
  });

  it('should support use aggregate transformation', () => {
    expect(Transformation.newTransformation().aggregate('Amount with sum as Total').toString()).toMatchSnapshot()
  });

  it('should support countSumPercents transformations', () => {
    expect(Transformation.newTransformation().topsum(10, 'Amount').toString()).toMatchSnapshot()
    expect(Transformation.newTransformation().topcount(15, 'Sales/Amount').toString()).toMatchSnapshot()
    expect(Transformation.newTransformation().toppercent(3, 'Age').toString()).toMatchSnapshot()
    expect(Transformation.newTransformation().bottomsum(10, 'Amount').toString()).toMatchSnapshot()
    expect(Transformation.newTransformation().bottomcount(15, 'Sales/Amount').toString()).toMatchSnapshot()
    expect(Transformation.newTransformation().bottompercent(3, 'Age').toString()).toMatchSnapshot()
  });

  it('should support identity tranformation', () => {
    expect(Transformation.newTransformation().identity().toString()).toMatchSnapshot()
  });

  it('should support concat transformation', () => {
    expect(
      Transformation
        .newTransformation()
        .concat(
          Transformation.newTransformation().topcount(2, 'Amount'),
          Transformation.newTransformation().aggregate('Amount'),
        )
        .toString()
    ).toMatchSnapshot()
  });

  it('should support filter transformation', () => {

    expect(transformation().filter(ODataFilter.newFilter().field('Amount').gt(3)).toString()).toMatchSnapshot()

  });

  it('should supoprt simply groupBy', () => {
    expect(
      Transformation.
        newTransformation().
        groupBy(['Customer/Country', 'Product/Name']).
        toString()
    )
      .toMatchSnapshot()

    expect(
      Transformation.
        newTransformation().
        groupBy(
          ['Customer/Country', 'Product/Name'],
          Transformation.newTransformation().aggregate('Amount with sum as Total')
        ).
        toString()
    ).toMatchSnapshot()
  });

});
