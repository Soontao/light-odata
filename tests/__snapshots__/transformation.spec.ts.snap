// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transofmration Test Suite should supoprt simply groupBy 1`] = `"groupBy((Customer/Country,Product/Name))"`;

exports[`Transofmration Test Suite should supoprt simply groupBy 2`] = `"groupBy((Customer/Country,Product/Name),aggregate(Amount with sum as Total))"`;

exports[`Transofmration Test Suite should support concat transformation 1`] = `""`;

exports[`Transofmration Test Suite should support countSumPercents transformations 1`] = `"topsum(10,Amount)"`;

exports[`Transofmration Test Suite should support countSumPercents transformations 2`] = `"topcount(15,Sales/Amount)"`;

exports[`Transofmration Test Suite should support countSumPercents transformations 3`] = `"toppercent(3,Age)"`;

exports[`Transofmration Test Suite should support countSumPercents transformations 4`] = `"bottomsum(10,Amount)"`;

exports[`Transofmration Test Suite should support countSumPercents transformations 5`] = `"bottomcount(15,Sales/Amount)"`;

exports[`Transofmration Test Suite should support countSumPercents transformations 6`] = `"bottompercent(3,Age)"`;

exports[`Transofmration Test Suite should support identity tranformation 1`] = `"identity"`;

exports[`Transofmration Test Suite should support use aggregate transformation 1`] = `"aggregate(Amount with sum as Total)"`;
